File: flutter_app/lib/repository/repository.dart
import 'package:flutter/widgets.dart';
import 'package:flutter_app/model/event.dart';
import 'package:flutter_app/repository/local_database.dart';
import 'package:flutter_app/repository/web_api.dart';

class Repository extends ChangeNotifier {
  final pageSize = 20;
  final WebApi webApi;
  final LocalDatabase localDatabase;

  List<Event> _events = [];
  int _currentPageNumber = 0;

  Repository(this.webApi, this.localDatabase) {
    loadMoreEvents();
  }

  List<Event> getEvents() {
    return _events;
  }

  Future<void> purge() async {
    await localDatabase.deleteEvents();
  }

  Future<void> loadMoreEvents() async {
    final loadedEvents = await _getNextEventsPage();
    _currentPageNumber++;
    if (loadedEvents.isNotEmpty) {
      await localDatabase.insertEvents(loadedEvents);
      await _refreshEvents();
    }
  }

  Future<List<Event>> _getNextEventsPage() {
    final nextPageNumber = _currentPageNumber + 1;
    return webApi.getEventsPage(nextPageNumber, pageSize);
  }

  Future<void> _refreshEvents() async {
    final eventsFromDb = await localDatabase.events();
    _events.clear();
    _events.addAll(eventsFromDb);
    notifyListeners();
  }
}

File: ios-app/ios-app/Repository/Repository.swift
//
//  Repository.swift
//  ios-app
//
//  Created by admin on 01/04/2021.
//

import Foundation

class Repository : RepositoryProtocol {
    private let webApi: WebApi
    private let localDatabase: LocalDatabase
    
    @Published var events: [Event]
    
    private var eventsPageNumber = 0
    
    init(webApi: WebApi, localDatabase: LocalDatabase) {
        self.webApi = webApi
        self.localDatabase = localDatabase
        events = []
        loadNextEventsPage()
    }
    
    func loadNextEventsPage() -> Void {
        webApi.getEventsPage(pageNumber: eventsPageNumber + 1, pageSize: 20) { loadedEvents in
            if !loadedEvents.isEmpty {
                self.localDatabase.insertEvents(events: loadedEvents)
                self.refreshEvents()
                self.eventsPageNumber += 1
            }
        }
    }
    
    private func refreshEvents() {
        let eventsFromDb = self.localDatabase.getAllEvents()
        DispatchQueue.main.async {
            self.events = eventsFromDb
        }
    }
}
